### PR TITLE
[Feature]: Awards Backend Schema

### PR DESCRIPTION
**Pull request by feng-eric**
Changes:
- Implemented award schema (many to many relationship with performance using awardperformance relation schema)
- Added relationship between award and user & award and ajudication
- Implemented CRUD operations for awards
- Added retrieving performances associated with an award and vice versa
- Reformated code structure in performances to improve readability (removed multiple nested if statements)
- Added retrieving school name for each performance
- Added setting userId for created adjudications and awards from session
- Added nominee_count to award performance table and unique constraint (award_id, performance_id)
- Created endpoint to finalize an award and endpoint to nominate a performance for an award
- Improved delete functionality to delete associated adjudications & award performances when deleting an award
- Improved update for performance and award to be able to update provided fields (no longer need to pass all fields)
- Added status to awardperformances table to track which performance is the finalist for the award

### CODE REVIEW

**Comment by thepenguinco** [comment]
If you haven't already taken a look here, you can use this to get the joins on relations: https://www.prisma.io/docs/concepts/components/prisma-schema/relations

**Comment by feng-eric** [comment]
> If you haven't already taken a look here, you can use this to get the joins on relations: https://www.prisma.io/docs/concepts/components/prisma-schema/relations

Yup, I did it based on the docs. It seems they use the relation table as the main table for querying

File: pages/api/adjudications/edit.js [hunk 1]
```diff
@@ -15,13 +15,13 @@ export default async (req, res) => {
...
       audioUrl,
       notes,
-      specialAward,
       performanceID,
     } = req.body;
 
     // If all params exist
     if (id && artisticMark && technicalMark && cumulativeMark && performanceID) {
       // Update adjudication
+      // TODO: set user_id from session???
```
Comment (feng-eric) [inline pages/api/adjudications/edit.js:24]: adjudication now should store data about the user who created the adjudication. Should I be getting the id from session or will it be passed from the frontend

Comment (thepenguinco) [reply pages/api/adjudications/edit.js:24]: Assume passed for now.  The frontend will/should know the user, user id, user name and user email.

File: prisma/schema.sql [hunk 2]
```diff
@@ -83,10 +83,31 @@ CREATE TABLE adjudications (
...
+  is_finalized BIT NOT NULL,
+  user_id INTEGER NOT NULL,
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  FOREIGN KEY(user_id) REFERENCES users(id)
+);
+
+-- Create awards_performances table
+CREATE TABLE awards_performances (
+  id SERIAL PRIMARY KEY NOT NULL,
```
Comment (feng-eric) [inline prisma/schema.sql:107]: TODO: add constraint to make (award_id, performance_id) unique

File: pages/api/adjudications/edit.js (hunk 1 above)
Comment (thepenguinco) [reply pages/api/adjudications/edit.js:24]: @feng-eric  actually, scrap that.  Let's use the session in the backend to find out all the user details.  I think that's probably better.

**Review by thepenguinco — APPROVED** [review]
Other than what we discussed about last time (addressing some TODOs like adding unique constraint), everything looks fine.

- Let's pull the user ID from the session.  There are multiple ways of doing this, let's just go with attaching the user ID to the session in NextAuth.  We can discuss in meeting.
- We'll need to standardize later error handling 
- Are there any other places we need to do basic error handling (i.e. deleting something and not cleaning up elsewhere will completely break the table schemas)
- Can we get some testing notes/documentation for how to call the specified endpoints maybe?

**Comment by feng-eric** [comment]
> * Are there any other places we need to do basic error handling (i.e. deleting something and not cleaning up elsewhere will completely break the table schemas)
When we delete a performance, we should delete all the performance details (adjudications) associated with it. When we delete an event, we should delete all performances associated with the event, and then all performance details associated with the performance. 
> * Can we get some testing notes/documentation for how to call the specified endpoints maybe?
Yeah, when all changes for awards are done, I'll start on documentation

File: pages/api/awards/finalize.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  // If it does not exist
+  if (!awardPerformance) {
+    return res.status(400).json({
+      error: 'Performance is not nominated for the provided award',
+    });
+  }
+
+  const isFinalized = awardPerformance.awards ? awardPerformance.awards.is_finalized : false;
+  if (isFinalized) {
```
Comment (feng-eric) [inline pages/api/awards/finalize.js:43]: If we can nominate multiple performances for the same award, I'll remove this

File: prisma/schema.sql [hunk 4]
```diff
@@ -83,11 +86,35 @@ CREATE TABLE adjudications (
...
+  user_id INTEGER NOT NULL,
   created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
   updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
-  FOREIGN KEY(performance_id) REFERENCES performances(id)
+  FOREIGN KEY(user_id) REFERENCES users(id)
+);
+
+-- Create awards_performances table
+CREATE TABLE awards_performances (
+  id SERIAL PRIMARY KEY NOT NULL,
```
Comment (feng-eric) [inline prisma/schema.sql:110]: Prisma's doc had this id in addition to the ids of the 2 associated tables in their documentation but it seems a bit redundant

File: pages/api/awards/collect.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import prisma from '@prisma/index'; // Prisma client
+import { getSession } from 'next-auth/client'; // Session handling
+
+// Get awards
+export default async (req, res) => {
+  // Collect session from request
+  const session = await getSession({ req });
+
+  //TODO: add functionality of getting awards by performance id
```
Comment (feng-eric) [inline pages/api/awards/collect.js:9]: Is this a use case we need

File: pages/api/awards/collect.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  // Collect all awards from database
+  const awards = await getAwards();
+
+  return res.json(awards);
+};
+
+export const getAwards = async () => {
+  // Collect awards with performanceID
+  // TODO: perhaps add flag to include the performances
```
Comment (feng-eric) [inline pages/api/awards/collect.js:24]: Do we need to retrieve awards for only certain performances

File: pages/api/performances/edit.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+  if (!id) {
+    return res.status(400).json({
+      error: 'Required performance id not provided',
+    });
+  }
+  console.log(danceTitle);
+
+  const editData = {};
+  if (name) editData.name = name;
```
Comment (feng-eric) [inline pages/api/performances/edit.js:37]: Is there a better way of doing this LOL

File: pages/api/awards/collect.js (hunk 5 above)
Comment (thepenguinco) [reply pages/api/awards/collect.js:9]: We should get awards by performance for feedback sharing.

File: pages/api/performances/edit.js (hunk 7 above)
Comment (thepenguinco) [reply pages/api/performances/edit.js:36]: for edit my impression was the frontend passes in all the data every time they call edit.  no need to support per field 'patching' I think.

File: pages/api/awards/collect.js (hunk 6 above)
Comment (thepenguinco) [reply pages/api/awards/collect.js:24]: What do you mean?

File: pages/api/awards/collect.js (hunk 5 above)
Comment (feng-eric) [reply pages/api/awards/collect.js:9]: This is already being done in feedback sharing :)

File: pages/api/awards/collect.js (hunk 6 above)
Comment (feng-eric) [reply pages/api/awards/collect.js:24]: Currently, when we retrieve an award, we always return the associated performance. Are there cases where we don't want to send this data or do we always want to send it

Comment (thepenguinco) [reply pages/api/awards/collect.js:24]: Let's always send it.  I see no reason why not to include it.

File: pages/api/awards/collect.js (hunk 5 above)
Comment (thepenguinco) [reply pages/api/awards/collect.js:9]: sounds good then we don't need this functionality for now.

File: prisma/schema.sql (hunk 4 above)
Comment (thepenguinco) [reply prisma/schema.sql:110]: think it's fine

**Pull request merged**